gtsam/geometry/Line3.h

```diff
@@ -0,0 +1,178 @@
+/* ----------------------------------------------------------------------------
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file    Line3.h
+ * @brief   4 dimensional manifold of 3D lines
+ * @author  Akshay Krishnan
+ * @author  Frank Dellaert
+ */
+// \callgraph
+
+#pragma once
+
+#include <gtsam/base/concepts.h>
+#include <gtsam/geometry/Rot3.h>
+#include <gtsam/geometry/SO3.h>
+#include <gtsam/nonlinear/expressions.h>
+#include <gtsam/slam/expressions.h>
+
+#include <boost/random.hpp>
+#include <cstdlib>
+#include <ctime>
+
+namespace gtsam{
+
+/**
+ * A 3D line (R,a,b) : (Rot3,Scalar,Scalar)
+ * @addtogroup geometry
+ * \nosubgrouping
+ */
+class Line3 {
+private:
+    Rot3 R_;    // Rotation of line about x and y in world frame
+    double a_, b_;  // Intersection of line with x-y world plane
+
+public:
+    enum {dimension = 4};
+
+    /** Default constructor is the Z axis **/
+    Line3() :
+            a_(0), b_(0)
+    {}
+
+    /** Constructor:
+     * Parallel to z axis, intersecting x-y plane at (a,b) **/
+    Line3(const double a, const double b):
+            a_(a), b_(b)
+    {}
+
+    /** Constructor for general line from (R, a, b) **/
+    Line3(const Rot3 &R, const double a, const double b) :
+            R_(R), a_(a), b_(b)
+    {}
+
+    /** Sample two points lying on the line
+     * @return Vector6 where first 3 elements are one point, and
+     * next 3 elements are another point **/
+    Vector6 points();
+
+    /**
+     * The retract method maps from the tangent space back to the manifold.
+     * The tangent space for the rotation of a line is only two dimensional -
+     * rotation about x and y
+     * @param v: increment in tangent space
+     * @param H: Jacobian of retraction with respect to the increment
+     * @return: resulting line after adding the increment and mapping to the manifold
+     */
+    Line3 retract(const Vector4 &v, OptionalJacobian<4,4> H = boost::none) const;
+
+    /**
+     * The localCoordinates method is the inverse of retract and finds the difference
+     * between two lines in the tangent space.
+     * @param q Line3 on manifold
+     * @param H OptionalJacobian of localCoordinates with respect to line
+     * @return Vector4 difference in the tangent space
+     */
+    Vector4 localCoordinates(const Line3 &q, OptionalJacobian<4,4> H = boost::none) const;
+
+    /**
+     * Rotation of line accessor
+     * @return Rot3
+     */
+    Rot3 R() const
+    {
+        return R_;
+    }
+
+    /**
+     * Accessor for a, b
+     * @return Vector2(a, b)
+     */
+    Vector2 V() const
```

Comment:
same

Reply:
Same as above, the accessor is necessary. 
Alternative names: translation() might be deceiving, but maybe closestPoint() or just point() is better?

Reply:
Some, make some friends. I really want to get rid of those accessors, as they break encapsulation. We should be able to change the representation of the line without anyone (except friends, who asked for it), noticing.